Source/Dafny/Verifier/ProofObligationDescription.cs

```diff
@@ -416,31 +416,34 @@ public class Terminates : ProofObligationDescription {
 
   public override string FailureDescription =>
     (inferredDescreases
-      ? "cannot prove termination; try supplying a decreases clause"
-      : "decreases clause might not decrease") +
+      ? ("cannot prove termination; try supplying a decreases clause" + (isLoop ? " for the loop" : ""))
+      : $"decreases {FormDescription} might not decrease") +
     (hint is null ? "" : $" ({hint})");
 
   public override string ShortDescription => "termination";
 
   private readonly bool inferredDescreases;
+  private readonly bool isLoop;
   private readonly string hint;
+  private string FormDescription => isLoop ? "expression" : "clause";
 
-  public Terminates(bool inferredDescreases, string hint = null) {
+  public Terminates(bool inferredDescreases, bool isLoop, string hint = null) {
     this.inferredDescreases = inferredDescreases;
+    this.isLoop = isLoop;
     this.hint = hint;
   }
 }
 
 public class DecreasesBoundedBelow : ProofObligationDescription {
   public override string SuccessDescription =>
-    $"decreases clause{component} is bounded below by {zeroStr}";
+    $"decreases {component} is bounded below by {zeroStr}";
 
   public override string FailureDescription =>
-    $"decreases clause{component} must be bounded below by {zeroStr}{suffix}";
+    $"decreases {component} must be bounded below by {zeroStr}{suffix}";
 
-  public override string ShortDescription => "bounded decreases clause";
+  public override string ShortDescription => "bounded decreases expression";
 
-  private string component => N == 1 ? "" : $" (component {k})";
+  private string component => N == 1 ? "expression" : $"clause (component at index {k})";
```

Comment:
```suggestion
  private string component => N == 1 ? "expression" : $"expression at index {k}";
```
As Rustan suggested, since we know the components index, we might as well use the shorter version of this error message.
I added another suggestion so that if you accept both, it should also fix the test.

Reply:
Ah, yes, I like this better, too.